LaTeX/LaTeX.sublime-syntax

```diff
@@ -198,6 +207,123 @@ contexts:
     - include: else-pop
 
+  xparse-argspec:
+    - match: (?=\{)
+      set:
+        # deliberately using an anonymous scope here. we need two match the `{`
+        # in two stages, so we can apply the `clear_scopes`, and still be able
+        # to push `xparse-argspec` onto the stack without messing up scoping.
+        - clear_scopes: 1
+        - match: '\{'
+          scope: punctuation.definition.group.brace.begin.latex
+          set: xparse-argspec-content
+
+    - include: paragraph-pop
+    - include: else-pop
+
+  xparse-argspec-content:
+    - clear_scopes: 1
+    - meta_scope: meta.function.parameters.latex
+    - include: brace-group-end
+    - include: macro-braces
+    - include: xparse-single-argspec
+    # not actually allowed, but included here to future-proof highlighting
+    - include: general-constants
+    - include: general-commands
+    - include: paragraph-pop
+
+  xparse-single-argspec:
+    # modifiers and simple specs
+    - match: '[+!movsb]'
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+
+    - match: t
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+      push:
+        - xparse-args-token
+
+    # delimited arguments without default
+    - match: '[rd]'
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+      push:
+        - xparse-args-token
+        - xparse-args-token
+
+    # delimited arguments with default
+    - match: '[RD]'
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+      push:
+        - xparse-optarg-value
+        - xparse-args-token
+        - xparse-args-token
+
+    # optional argument with default
+    - match: O
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+      push: xparse-optarg-value
+
+    # embellishment argument
+    - match: e
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+      push: xparse-embellishment
+
+    # embellishment argument with default
+    - match: E
+      scope: string.unquoted.latex meta.annotation.parameters.latex
+      push:
+        - xparse-optarg-value
+        - xparse-embellishment
+
+  xparse-embellishment:
+    - match: \{
+      scope: punctuation.definition.group.brace.begin.latex
+      set: xparse-embellishment-content
+    - match: (?=\S)
+      set: xparse-args-token
+    - include: paragraph-pop
+
+  xparse-embellishment-content:
+    - match: (?!\})
+      push: xparse-args-token
+    - include: brace-group-end
+    - include: paragraph-pop
+    - include: else-pop
+
+  xparse-args-token:
+    # either a command
+    - match: '(\\){{cmdname}}{{endcs}}'
+      scope: constant.other.token.latex
+      captures:
+        1: punctuation.definition.backslash.latex
+      pop: 1
+    # or the token wrapped in a group
+    - match: \{
+      scope: punctuation.definition.group.brace.begin.tex
+      push: xparse-args-token
+    - match: \}
+      scope: punctuation.definition.group.brace.end.tex
+      pop: 1
```

Comment:
With a look how this context is used, I have a feeling this could cause unbalanced closing braces to be consumed by accident.

Reply:
These lines result in function parameters not beeing terminated in case a stray `}` appears after a type specifier. 

While it would be ok in:

```latex
\NewDocumentCommand {\conjugate} { t } O{#1ed{}} O{#2} } {(#1,#2,#3)}
%                                ^^^^^^^^^^^^^^^^^^^^^^^ meta.function.parameters
```

... function body would be turned into parameters in scenarios, like:

```latex
\NewDocumentCommand {\conjugate} { t } {(#1,#2,#3)}
%                                ^^^^^^^^^^^^^^^^^^ meta.function.parameters
```

None of them is ideal, but I tend to recommend to prefer correctly stacked braced contexts. It has proven robust in other syntaxes of complex code architectures at least.

```latex
\NewDocumentCommand {\conjugate} { t } {(#1,#2,#3)}
%                                ^^^^^ meta.function.parameters
%                                     ^^^^^^^^^^^^^ - meta.function.parameters
```

Reply:
you're right. should we mark the offending brace as `invalid`, or just skip out of the context?

Reply:
Fail early and skip the rest is often the best and most general approach. It's very likely to have incomplete code while editing. Too many invalid scopings may quickly become annoying.